### PR TITLE
Migrate to python3

### PR DESCRIPTION
**Pull request by joloh**
Migrate to python3

### CODE REVIEW

File: debian/changelog [hunk 1]
```diff
@@ -1,3 +1,10 @@
+tiqit (1.1.0-1) trusty; urgency=low
+
+  * Move to python3
+  * Remove browser check
```
Comment (matthewhall79) [inline debian/changelog]: Add a few more words so it's clearer what check this is referring to. (And add full stops to both bullets.)

**Review by matthewhall79 — CHANGES_REQUESTED** [review]

File: debian/changelog (hunk 1 above)
Comment (joloh) [reply debian/changelog]: Done

**Review by matthewhall79 — APPROVED** [review]

File: api/api.py [hunk 2]
```diff
@@ -27,25 +27,27 @@
...
 
 from tiqit import *
-import os, re, sys, urllib, codecs, commands, locale
+import os, re, sys, urllib.request, urllib.parse, urllib.error, codecs, subprocess, locale
 
 # Set environment variables so that processes called by Tiqit interpret
 # decode/encode input/output as UTF-8.
-for envvar in ['LC_ALL', 'LANG', 'LANGUAGE']:
-    os.environ[envvar] = 'en_US.UTF-8'
+# @@@ JL: Need to stop output as bytes?
```
Comment (ollij93) [inline api/api.py]: Flagging @@@'s here - maybe something for discussion?

File: make_deb [hunk 3]
```diff
@@ -88,8 +89,9 @@ mkdir_or_die "$DOC_DIR/examples"
 # Put files into $MAIN_DIR
 # Note this excludes tiqit.db, tiqit.pickle and fielddata.js: These will be
 # created by the post-install script (@@@ TBD).
```
Comment (ollij93) [inline make_deb]: Comment could do with updating

File: scripts/actions/bugbox.py [hunk 4]
```diff
@@ -37,7 +37,7 @@
 if len(bugids) == 0:
     # Look for a Named Bug name in the bugbox.
     match = []
-    for bug in prefs.ofType('namedBug').keys():
+    for bug in list(prefs.ofType('namedBug').keys()):
```
Comment (ollij93) [inline scripts/actions/bugbox.py]:
```suggestion
    for bug in prefs.ofType('namedBug'):
```

File: scripts/actions/bugbox.py [hunk 5]
```diff
@@ -50,10 +50,10 @@
             match = match[0]
 
         sendMessage(MSG_INFO, "Redirected to named bug.")
-        redirect('view/%s' % urllib.quote(match[8:]))
+        redirect('view/%s' % urllib.parse.quote(match[8:]))
 
     # Look for a named query
-    for search in prefs.ofType('search').keys():
+    for search in list(prefs.ofType('search').keys()):
```
Comment (ollij93) [inline scripts/actions/bugbox.py]:
```suggestion
    for search in prefs.ofType('search'):
```

File: scripts/actions/edit.py [hunk 6]
```diff
@@ -38,12 +38,12 @@
...
 fieldsInUpdate = [x.name for x in fieldsInUpdate if x.editable]
 
-if args.has_key('S1S2-without-workaround'):
+if 'S1S2-without-workaround' in args:
     fieldsInUpdate.append('S1S2-without-workaround')
 
 # Make checkboxes into Y's
 for field in [x for x in fieldsInUpdate if allFields[x].type == 'Boolean']:
-    if args.has_key(field) and args[field] and args[field] != 'false':
+    if field in args and args[field] and args[field] != 'false':
```
Comment (ollij93) [inline scripts/actions/edit.py]:
```suggestion
    if args.get(field) and args.get(field) != 'false':
```

File: scripts/actions/edit.py [hunk 7]
```diff
@@ -65,7 +65,7 @@
     changes[field] = fieldObj.filterEdit(args, args[field])
 
 # Now fix multi value fields
-for field in changes.keys():
+for field in list(changes.keys()):
```
Comment (ollij93) [inline scripts/actions/edit.py]:
```suggestion
for field in changes:
```

File: scripts/actions/edit.py [hunk 8]
```diff
@@ -74,7 +74,7 @@
 # the browser but the backend will not let it be updated. This is true
 # for e.g. greyed out fields like forwarded to
 newData = OverriddenData(data, changes)
-for field in changes.keys():
+for field in list(changes.keys()):
     if not isValidField(allFields[field], newData):
         del changes[field]
```
Comment (ollij93) [inline scripts/actions/edit.py]: Better written as something like this:
```suggestion
changes = {
    field: value
    for field, value in changes.items()
    if isValidField(allFields[field], newData)
}
```

File: scripts/actions/edit.py [hunk 9]
```diff
@@ -103,14 +103,14 @@
...
-if args.has_key('newRelates'):
+if 'newRelates' in args:
     changes['Related-bugs'] = extractBugIds(args['newRelates'])
     if not changes['Related-bugs']:
         del changes['Related-bugs']
 
 bugView.prepareUpdateBug(changes, fieldsInUpdate, data)
 
-changes_save = dict([(allFields[field].savename, changes[field]) for field in changes.keys()])
+changes_save = dict([(allFields[field].savename, changes[field]) for field in list(changes.keys())])
```
Comment (ollij93) [inline scripts/actions/edit.py]:
```suggestion
changes_save = dict([(allFields[field].savename, changes[field]) for field in changes])
```

File: scripts/actions/saveprefs.py [hunk 10]
```diff
@@ -10,7 +10,7 @@
 
 defsPerBugView = copy.deepcopy(Prefs.defaults['miscDefaultsPerBugView'])
 
-for f in fields.keys():
+for f in list(fields.keys()):
```
Comment (ollij93) [inline scripts/actions/saveprefs.py]:
```suggestion
for f in fields:
```

File: scripts/actions/submit.py [hunk 11]
```diff
@@ -25,20 +25,20 @@
...
-    if args.has_key(field) and args[field] != 'N':
+    if field in args and args[field] != 'N':
         args[field] = 'Y'
     else:
         args[field] = 'N'
 
 # Convert values if appropriate
-for arg in args.keys():
-    if allFields.has_key(arg):
+for arg in list(args.keys()):
```
Comment (ollij93) [inline scripts/actions/submit.py]:
```suggestion
for arg in args:
```

File: scripts/backend/__init__.py [hunk 12]
```diff
@@ -29,15 +29,15 @@
...
+from io import StringIO
 
 _backends = {}
 def initialise():
     global allFields
     allFields = fields.load()
 
     # Import all the configured backends
-    for prefix, mod in Config().section('backends').items():
+    for prefix, mod in list(Config().section('backends').items()):
```
Comment (ollij93) [inline scripts/backend/__init__.py]: I assume this is 2to3 that has been wrapping all these things in lists?
Seems like an odd thing for it to do within a for loop, but maybe it's not that smart.
```suggestion
    for prefix, mod in Config().section('backends').items():
```

File: scripts/fields.py [hunk 13]
```diff
@@ -264,7 +264,7 @@ def hasParentDependency(self):
     def getAllValues(self):
         if self.hasParentDependency():
             vals = []
-            for more in self._perParentFieldValues.values():
+            for more in list(self._perParentFieldValues.values()):
```
Comment (ollij93) [inline scripts/fields.py]:
```suggestion
            for more in self._perParentFieldValues.values():
```

File: scripts/fields.py [hunk 14]
```diff
@@ -316,15 +316,15 @@ def getOptions(vals, descs):
                     return ["%s - %s" % (v, d) if v != "" and d != "" else "" for v, d in zip(vals, descs)]
                 a = dict((k, getOptions(self._perParentFieldValues[k],
                                         self._perParentFieldDescs[k])) for k in
-                                self._perParentFieldDescs.keys())
+                                list(self._perParentFieldDescs.keys()))
```
Comment (ollij93) [inline scripts/fields.py]:
```suggestion
                               self._perParentFieldDescs.keys())
```

File: scripts/generatejs.py [hunk 15]
```diff
@@ -37,20 +37,20 @@ def __init__(self):
        self.interns = {}
        self.counts = {}
    def intern(self):
-       prio = [(v, k) for k, v in self.counts.items()]
+       prio = [(v, k) for k, v in list(self.counts.items())]
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
       prio = [(v, k) for k, v in self.counts.items()]
```

File: scripts/generatejs.py [hunk 16]
```diff
@@ -61,7 +61,7 @@ def __getitem__(self, key):
 I = Intern()
 
 def makeInterns(out):
-    for f in backend.allFields.values():
+    for f in list(backend.allFields.values()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
    for f in backend.allFields.values():
```

File: scripts/generatejs.py [hunk 17]
```diff
@@ -86,7 +86,7 @@ def makeInterns(out):
                 for d in f.descs:
                     I.count(utils.encodeJS(d))
         elif f._perParentFieldValues:
-            for key in f._perParentFieldValues.keys():
+            for key in list(f._perParentFieldValues.keys()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
            for key in f._perParentFieldValues:
```

File: scripts/generatejs.py [hunk 18]
```diff
@@ -108,12 +108,12 @@ def makeInterns(out):
             for i in f.defaultsFor:
                 I.count(utils.encodeJS(i))
         if f._mandatoryIf:
-            for k, v in f._mandatoryIf.items():
+            for k, v in list(f._mandatoryIf.items()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
            for k, v in f._mandatoryIf.items():
```

File: scripts/generatejs.py [hunk 19]
```diff
@@ -108,12 +108,12 @@ def makeInterns(out):
...
                 I.count(utils.encodeJS(k))
                 for i in v:
                     I.count(utils.encodeJS(i))
         if f._bannedIf:
-            for k, v in f._bannedIf.items():
+            for k, v in list(f._bannedIf.items()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
            for k, v in f._bannedIf.items():
```

File: scripts/generatejs.py [hunk 20]
```diff
@@ -126,7 +126,7 @@ def makeInterns(out):
                     I.count(utils.encodeJS(i))
 
     I.intern()
-    for k, v in I.interns.items():
+    for k, v in list(I.interns.items()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
    for k, v in I.interns.items():
```

File: scripts/generatejs.py [hunk 21]
```diff
@@ -140,7 +140,7 @@ def JSListOfListsFromSequence(arg):
     return "[%s]" % ",".join(map(JSListFromSequence, arg))
 
 def genFieldLists(out):
-    for f in backend.allFields.values():
+    for f in list(backend.allFields.values()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
    for f in backend.allFields.values():
```

File: scripts/generatejs.py [hunk 22]
```diff
@@ -150,30 +150,30 @@ def genFieldLists(out):
 def genRelationships(out):
     # Calculate reverseDefaultsWith first.
     reverseDefaultsWith = {}
-    for f in backend.allFields.values():
+    for f in list(backend.allFields.values()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
    for f in backend.allFields.values():
```

File: scripts/generatejs.py [hunk 23]
```diff
@@ -150,30 +150,30 @@ def genFieldLists(out):
...
         for f2name in f.defaultsWith:
             reverseDefaultsWith.setdefault(f2name, set()).add(f.name)
 
     # Only generates rels and vals for ones with multiple options
-    for f in backend.allFields.values():
+    for f in list(backend.allFields.values()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
    for f in backend.allFields.values():
```

File: scripts/generatejs.py [hunk 24]
```diff
@@ -150,30 +150,30 @@ def genFieldLists(out):
...
+                vals = list(map(list, list(zip(f.values, getOptions(f.values, f.descs)))))
             else:
-                vals = map(list, zip(f.values, f.descs))
+                vals = list(map(list, list(zip(f.values, f.descs))))
             out.write("""I.values=%s;\n""" % JSListOfListsFromSequence(vals))
             rewriterels = True
         elif f._perParentFieldValues:
             out.write("P=I.perparentvalues;\n")
-            for key in f._perParentFieldValues.keys():
+            for key in list(f._perParentFieldValues.keys()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
            for key in f._perParentFieldValues:
```

File: scripts/generatejs.py [hunk 25]
```diff
@@ -150,30 +150,30 @@ def genFieldLists(out):
...
-    for f in backend.allFields.values():
+    for f in list(backend.allFields.values()):
         out.write("I=F[%s];\n" % encodeJS(f.name))
         rewriterels = False
         def getOptions(values, descs):
             return ("%s - %s" % (v,d) if v else v for v,d in zip(values,descs))
         if f.values:
             if f.descs != f.values:
-                vals = map(list, zip(f.values, getOptions(f.values, f.descs)))
+                vals = list(map(list, list(zip(f.values, getOptions(f.values, f.descs)))))
```
Comment (ollij93) [inline scripts/generatejs.py]: I feel like a lot of these `map` calls should be re-written in a more pythonic way, but TBH I'm struggling to parse what they are doing :/

Perhaps using list comprehensions is the correct python3 way to do this?

File: scripts/generatejs.py [hunk 26]
```diff
@@ -190,11 +190,11 @@ def getOptions(values, descs):
             out.write("""I.reverseDefaultsWith=%s;\n""" % JSListFromSequence(reverseDefaultsWith[f.name]))
         if f._mandatoryIf:
             out.write("M=I.mandatoryif;\n")
-        for key in f._mandatoryIf.keys():
+        for key in list(f._mandatoryIf.keys()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
        for key in f._mandatoryIf:
```

File: scripts/generatejs.py [hunk 27]
```diff
@@ -190,11 +190,11 @@ def getOptions(values, descs):
...
             out.write("""M[%s]=%s;\n""" % (encodeJS(key), JSListFromSequence(f._mandatoryIf[key])))
         if f._bannedIf:
             out.write("B=I.bannedif;\n")
-        for key in f._bannedIf.keys():
+        for key in list(f._bannedIf.keys()):
```
Comment (ollij93) [inline scripts/generatejs.py]:
```suggestion
        for key in f._bannedIf:
```

File: scripts/helpers/lookup.py [hunk 28]
```diff
@@ -38,17 +38,17 @@ def getDefaultsForField(field):
...
-    print "<defaultfrom name='%s'>" % key.replace("'", "&apos;")
+    print("<defaultfrom name='%s'>" % key.replace("'", "&apos;"))
     for entry in revs[key]:
-        print "<entry>"
-        for k, v in entry.items():
-            print "<field name='%s'><![CDATA[%s]]></field>" % (k.replace("'", '&apos;'), encodeCDATA(v))
-        print "</entry>"
-    print "</defaultfrom>"
+        print("<entry>")
+        for k, v in list(entry.items()):
```
Comment (ollij93) [inline scripts/helpers/lookup.py]:
```suggestion
        for k, v in entry.items():
```

File: scripts/helpers/updateprefs.py [hunk 29]
```diff
@@ -13,7 +13,7 @@
 # Map arguments into a set() of deletes, and a dict() of sets.
 # A deleted item has the form "key<N>=<prefName>".
 # A changed item has the form "key<N>=<prefName>&val<N>=<prefVal>"
-for kArg in [x for x in args.keys() if x.startswith("key")]:
+for kArg in [x for x in list(args.keys()) if x.startswith("key")]:
```
Comment (ollij93) [inline scripts/helpers/updateprefs.py]:
```suggestion
for kArg in [x for x in args if x.startswith("key")]:
```

File: scripts/helpers/userdetails.py [hunk 30]
```diff
@@ -22,28 +22,28 @@
...
+results = list(mapnames.values())
 scope = ldap.SCOPE_SUBTREE
 
 query = num and '%s=%d' % (mapnames['unum'], num) or '%s=%s' % (mapnames['uname'], name)
 
 l = ldap.open(cfg.get('ldap_server'))
 r = l.search(base, scope, query, results)
 
-info = dict(zip(props, (name, num, 'Unknown', 'Unknown', '', '', '', '', 0)))
+info = dict(list(zip(props, (name, num, 'Unknown', 'Unknown', '', '', '', '', 0))))
```
Comment (ollij93) [inline scripts/helpers/userdetails.py]: `dict(list(...))`

😳 wat?

Comment (ollij93) [reply scripts/helpers/userdetails.py]: Maybe a dict comprehension like this does the same?
```suggestion
info = {k:v for k,v in zip(props, (name, num, 'Unknown', 'Unknown', '', '', '', '', 0))}
```

File: scripts/helpers/userdetails.py [hunk 31]
```diff
@@ -22,28 +22,28 @@
...
-if info.has_key('phone2') and not info['phone']:
+if 'phone2' in info and not info['phone']:
     info['phone'] = info['phone2']
 
 printXMLPageHeader()
 printXMLSectionHeader("userdetails")
 
-for key, val in info.items():
-    print "    <%s><![CDATA[%s]]></%s>" % (key, encodeCDATA(str(val)), key)
+for key, val in list(info.items()):
```
Comment (ollij93) [inline scripts/helpers/userdetails.py]:
```suggestion
for key, val in info.items():
```

File: scripts/index.py [hunk 32]
```diff
@@ -25,28 +25,30 @@
...
 
 from tiqit import *
-import os, re, sys, urllib, codecs, commands, locale
+import os, re, sys, urllib.request, urllib.parse, urllib.error, codecs, subprocess, locale
 
 # Set environment variables so that processes called by Tiqit interpret
 # decode/encode input/output as UTF-8.
-for envvar in ['LC_ALL', 'LANG', 'LANGUAGE']:
-    os.environ[envvar] = 'en_US.UTF-8'
+# @@@ JL: Need to stop output as bytes?
```
Comment (ollij93) [inline scripts/index.py]: More @@@s, same as before :)

File: scripts/index.py [hunk 33]
```diff
@@ -164,7 +166,7 @@ def addSlash():
     # Resplit the path to to merge superfluous arguments, to allow for example
     # looking up a "TCP/IP" named search where the "/" is unencoded.
     path = os.environ['PATH_INFO'].lstrip('/').split('/', len(page[1]) - 1)
-    args = '&'.join(map('='.join, zip(page[1], path)))
+    args = '&'.join(map('='.join, list(zip(page[1], path))))
```
Comment (ollij93) [inline scripts/index.py]: Think I saw this elsewhere as well.
Could a list comprehension like this do the same thing?
`'&'.join(f"{k}={v}" for k,v in zip(page[1], path))`

File: scripts/loadclasses.py [hunk 34]
```diff
@@ -37,7 +37,7 @@ def _main():
     tiqit.database.initialise()
     tiqit.database.set('tiqit.classes', classes)
     tiqit.database.set('tiqit.projmap', projmap)
-    for key, val in other.items():
+    for key, val in list(other.items()):
```
Comment (ollij93) [inline scripts/loadclasses.py]:
```suggestion
    for key, val in other.items():
```

File: scripts/loadfields.py [hunk 35]
```diff
@@ -14,36 +14,37 @@
 import generatejs
 import tiqit
 import tiqit.database
+import sys
 
 
 def _save_fields(fields):
     # Before saving, we want to intern all the strings we can! This is to
     # reduce the size of tiqit.pickle.
```
Comment (ollij93) [inline scripts/loadfields.py]: Can we check if this actually saves any space and ditch this if it's minimal please?
It just seems like an area that's bound to cause trouble at some point.

Comment (ollij93) [reply scripts/loadfields.py]: FWIW I think python3 is much better about doing sensible amounts of this kind of thing itself, so expect this does very little.

File: scripts/pages/defaultvals.py [hunk 36]
```diff
@@ -30,23 +30,23 @@ def showFieldChoices():
     printPageHeader(PAGE_DEFVALS, "Default Field Values")
     printMessages()
 
-    fields = [f for f in allFields.values() if f.defaultsWith]
+    fields = [f for f in list(allFields.values()) if f.defaultsWith]
```
Comment (ollij93) [inline scripts/pages/defaultvals.py]:
```suggestion
    fields = [f for f in allFields.values() if f.defaultsWith]
```

File: scripts/pages/env.py [hunk 37]
```diff
@@ -22,8 +22,8 @@
...
 import os
 
-print "Content-Type: text/plain"
-print
+print("Content-Type: text/plain")
+print()
 
-for f in os.environ.keys():
-    print "%s = %s" % (f, os.environ[f])
+for f in list(os.environ.keys()):
```
Comment (ollij93) [inline scripts/pages/env.py]:
```suggestion
for f in os.environ:
```

File: scripts/pages/fields.py [hunk 38]
```diff
@@ -28,9 +28,9 @@
 printPageHeader(PAGE_FIELDS, hideSavedSearches=True, hideNamedBugs=True)
 printMessages()
 
-print "<h1>Fields</h1>"
+print("<h1>Fields</h1>")
 
-fields = allFields.keys()
+fields = list(allFields.keys())
 fields.sort()
```
Comment (ollij93) [inline scripts/pages/fields.py]:
```suggestion
fields = sorted(allFields.keys())
```

File: scripts/pages/results.py [hunk 39]
```diff
@@ -73,7 +73,7 @@
     for parent in field._parentFields:
         parents.add(allFields[parent])
 
-    for bannedIf in field._bannedIf.keys():
+    for bannedIf in list(field._bannedIf.keys()):
```
Comment (ollij93) [inline scripts/pages/results.py]:
```suggestion
    for bannedIf in field._bannedIf:
```

File: scripts/pages/results.py [hunk 40]
```diff
@@ -241,29 +241,29 @@ def printTableFooter():
         # attributes whether their values are empty or not
         #
         attr_values = ["'%s'" % encodeHTML(bug[x.name]) for x in parents]
-        attributes = map("=".join, zip(attr_names, attr_values))
+        attributes = list(map("=".join, list(zip(attr_names, attr_values))))
```
Comment (ollij93) [inline scripts/pages/results.py]:
```suggestion
        attributes = [f"{name}={val}" for name,val in zip(attr_names, attr_values)]
```

File: scripts/tiqit/__init__.py [hunk 41]
```diff
@@ -87,7 +87,7 @@ def extractBugIds(string):
     normalise tqtSd12345 into TQTsd12345).
     """
 
-    backends = (k for k, v in Config().section('backends').items())
+    backends = (k for k, v in list(Config().section('backends').items()))
```
Comment (ollij93) [inline scripts/tiqit/__init__.py]:
```suggestion
    backends = (k for k, v in Config().section('backends').items())
```

File: scripts/tiqit/__init__.py [hunk 42]
```diff
@@ -131,29 +131,34 @@ def __init__(self):
...
     def __getitem__(self, key):
-        val = ''
-        if self.overrides.has_key(key):
+        if key in self.overrides:
+            #val = self.overrides[key].encode('utf-8')
             val = self.overrides[key]
-        elif self.cgi.has_key(key):
-            val = self.cgi[key].value.decode('utf-8')
+        elif key in self.cgi:
+            # @@@ JL: Need to stop output as bytes?
```
Comment (ollij93) [inline scripts/tiqit/__init__.py]: More @@@s :)

File: scripts/tiqit/__init__.py [hunk 43]
```diff
@@ -237,31 +242,36 @@ class ConfigSection:
...
-                return []
-            return map(string.strip, re.compile(r'[\n,]').split(self[key]))
+            ret_list = self[key].split()
+            return ret_list if ret_list else []                
         def items(self, raw=False):
             items = self.cfg.items(self.name, raw)
-            opts = Config.cfg_defaults.has_key(self.name) and Config.cfg_defaults[self.name] or {}
-            items.extend([(x, opts[x]) for x in opts.keys() if not self.cfg.has_option(self.name, x)])
+            opts = self.name in Config.cfg_defaults and Config.cfg_defaults[self.name] or {}
+            items.extend([(x, opts[x]) for x in list(opts.keys()) if not self.cfg.has_option(self.name, x)])
```
Comment (ollij93) [inline scripts/tiqit/__init__.py]:
```suggestion
            items.extend([(x, opts[x]) for x in opts if not self.cfg.has_option(self.name, x)])
```

File: scripts/tiqit/__init__.py [hunk 44]
```diff
@@ -288,8 +298,8 @@ def readConfig(self, paths):
             raise Exception('Config file not found in any of the expected ' +
                     'locations: %s' % ', '.join(CFG_PATHS))
 
-        self.cfg = ConfigParser.ConfigParser()
-        [self.cfg.add_section(sect) for sect in Config.cfg_defaults.keys()]
+        self.cfg = configparser.ConfigParser()
+        [self.cfg.add_section(sect) for sect in list(Config.cfg_defaults.keys())]
```
Comment (ollij93) [inline scripts/tiqit/__init__.py]:
```suggestion
        [self.cfg.add_section(sect) for sect in Config.cfg_defaults]
```

File: scripts/tiqit/__init__.py [hunk 45]
```diff
@@ -351,7 +361,7 @@ def generateToken():
     return token
 
 def authenticate():
-    if not os.environ.has_key('REMOTE_USER') or not os.environ['REMOTE_USER']:
+    if 'REMOTE_USER' not in os.environ or not os.environ['REMOTE_USER']:
```
Comment (ollij93) [inline scripts/tiqit/__init__.py]:
```suggestion
    if not os.environ.get('REMOTE_USER'):
```

File: scripts/tiqit/__init__.py [hunk 46]
```diff
@@ -587,10 +575,10 @@ def printMessages():
 
     # Print out any message in a cookie
     try:
-        if os.environ.has_key('HTTP_COOKIE'):
+        if 'HTTP_COOKIE' in os.environ:
             for cookie in os.environ['HTTP_COOKIE'].split(';'):
-                name, update = map(urllib.unquote,
-                                   cookie.strip().split('=', 1))
+                name, update = list(map(urllib.parse.unquote,
+                                   cookie.strip().split('=', 1)))
```
Comment (ollij93) [inline scripts/tiqit/__init__.py]:
```suggestion
                name, update = (urllib.parse.unquote(x) for x in cookie.strip().split('=', 1))
```

File: scripts/tiqit/__init__.py [hunk 47]
```diff
@@ -847,7 +835,7 @@ def _loadNamedPage(ofType):
         if not arg.find('=') >= 0:
             break
         key, val = arg.split('=', 1)
-        key, val = map(urllib.unquote, (key, val))
+        key, val = list(map(urllib.parse.unquote, (key, val)))
```
Comment (ollij93) [inline scripts/tiqit/__init__.py]: Won't let me provide the suggestion: But can do on a single line and avoid using `map` as above

File: scripts/tiqit/database.py [hunk 48]
```diff
@@ -32,7 +32,7 @@
 
 import tiqit
 import sqlite3 as sqlite, os.path
-import cPickle as pickle
+import pickle as pickle
```
Comment (ollij93) [inline scripts/tiqit/database.py]:
```suggestion
import pickle
```

Comment (ollij93) [reply scripts/tiqit/database.py]: Worth checking if this has happened in any other files as I may not have spotted it.

File: scripts/tiqit/defaults.py [hunk 49]
```diff
@@ -22,7 +22,7 @@ def _encodeValue(field, data):
     return '&'.join(map(encodeHTML, vals))
 
 def _decodeValue(field, value):
-    return dict(zip(field.defaultsWith, map(decodeHTML, value.split('&'))))
+    return dict(list(zip(field.defaultsWith, list(map(decodeHTML, value.split('&'))))))
```
Comment (ollij93) [inline scripts/tiqit/defaults.py]: Ergh, another `dict(list(...))` thing that should probably be a comprehension of some form

File: scripts/tiqit/defaults.py [hunk 50]
```diff
@@ -127,8 +127,8 @@ def _saveDefaultsInternal(field, data, defs, cu):
 
     # First throw out any defs that are already in the DB (or aren't in the DB
     # fo delete operations).
-    newdefs = dict((k, v) for k, v in defs.iteritems() if v)
-    for key, val in defs.items():
+    newdefs = dict((k, v) for k, v in defs.items() if v)
+    for key, val in list(defs.items()):
```
Comment (ollij93) [inline scripts/tiqit/defaults.py]:
```suggestion
    for key, val in defs.items():
```

File: scripts/tiqit/defaults.py [hunk 51]
```diff
@@ -140,7 +140,7 @@ def _saveDefaultsInternal(field, data, defs, cu):
     defs = newdefs
                 
     # Do the update.
-    for key, val in defs.items():
+    for key, val in list(defs.items()):
```
Comment (ollij93) [inline scripts/tiqit/defaults.py]:
```suggestion
    for key, val in defs.items():
```

File: scripts/tiqit/queries.py [hunk 52]
```diff
@@ -136,7 +136,7 @@ def parseQuery(args):
         # Check for special fields.
         # The only generic special field is dates (soon to be made even better!)
         if allFields[field].type == 'Date':
-            day, mon, year = map(int, val.split('/'))
+            day, mon, year = list(map(int, val.split('/')))
```
Comment (ollij93) [inline scripts/tiqit/queries.py]:
```suggestion
            day, mon, year = (int(x) for x in val.split('/'))
```

File: scripts/util/convert.py [hunk 53]
```diff
@@ -68,8 +68,8 @@ def readArgs():
 
 newPrefs = {}
 
-for p in oldPrefs.keys():
-  print "Copying profile for", p
+for p in list(oldPrefs.keys()):
```
Comment (ollij93) [inline scripts/util/convert.py]:
```suggestion
for p in oldPrefs:
```

File: scripts/utils.py [hunk 54]
```diff
@@ -88,17 +88,17 @@ def stringToColour(string):
...
+    os.write(fileObj, contents.encode(encoding))
     os.close(fileObj)
 
     return filename
 
 def memoize(f):
     cache = {}
     def memoized(*args, **kwargs):
-        key = (args, frozenset(kwargs.items()))
+        key = (args, frozenset(list(kwargs.items())))
```
Comment (ollij93) [inline scripts/utils.py]:
```suggestion
        key = (args, frozenset(kwargs.items()))
```

File: tiqitlib/tiqitlib.py [hunk 55]
```diff
@@ -144,7 +144,7 @@ def parse_args(siteurl):
 
 def process_cli_args(args):
     pp = pprint.PrettyPrinter(indent=4)
-    print ""
+    print("")
```
Comment (ollij93) [inline tiqitlib/tiqitlib.py]:
```suggestion
    print()
```

**Review by ollij93 — CHANGES_REQUESTED** [review]

File: api/api.py (hunk 2 above)
Comment (joloh) [reply api/api.py]: nah was more of a committed note for myself. have removed.

File: scripts/generatejs.py (hunk 25 above)
Comment (joloh) [reply scripts/generatejs.py]: I've attempted to clean them up. Can you have a look ? @ollij93 https://github.com/ensoft/tiqit/pull/56/commits/e3f6f0e0fe15bf2a90c1585aadbf51b4881c47c0

**Review by ollij93 — APPROVED** [review]

**Pull request merged**